pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java

```diff
@@ -18,139 +18,254 @@
-    private ElasticSearchConfig elasticSearchConfig;
+    private ElasticsearchConfig elasticSearchConfig;
+    private ElasticsearchClient elasticsearchClient;
+    private ObjectMapper objectMapper = new ObjectMapper();
 
     @Override
     public void open(Map<String, Object> config, SinkContext sinkContext) throws Exception {
-        elasticSearchConfig = ElasticSearchConfig.load(config);
+        elasticSearchConfig = ElasticsearchConfig.load(config);
         elasticSearchConfig.validate();
-        createIndexIfNeeded();
+        elasticsearchClient = new ElasticsearchClient(elasticSearchConfig);
     }
 
     @Override
-    public void close() throws Exception {
-        client.close();
+    public void close() {
+        if (elasticsearchClient != null) {
+            elasticsearchClient.close();
+            elasticsearchClient = null;
+        }
     }
 
     @Override
-    public void write(Record<byte[]> record) {
-        KeyValue<String, byte[]> keyValue = extractKeyValue(record);
-        IndexRequest indexRequest = Requests.indexRequest(elasticSearchConfig.getIndexName());
-        indexRequest.type(elasticSearchConfig.getTypeName());
-        indexRequest.source(keyValue.getValue(), XContentType.JSON);
-
-        try {
-        IndexResponse indexResponse = getClient().index(indexRequest, RequestOptions.DEFAULT);
-            if (indexResponse.getResult().equals(DocWriteResponse.Result.CREATED)) {
-                record.ack();
-            } else {
-                record.fail();
+    public void write(Record<GenericObject> record) throws Exception {
+        if (!elasticsearchClient.isFailed()) {
+            try {
+                Pair<String, String> idAndDoc = extractIdAndDocument(record);
+                if (idAndDoc.getRight() == null) {
+                    switch (elasticSearchConfig.getNullValueAction()) {
+                        case DELETE:
+                            if (elasticSearchConfig.isBulkEnabled()) {
+                                elasticsearchClient.bulkDelete(record, idAndDoc.getLeft());
+                            } else {
+                                elasticsearchClient.deleteDocument(record, idAndDoc.getLeft());
+                            }
+                            break;
+                        case IGNORE:
+                            break;
+                        case FAIL:
+                            elasticsearchClient.failed(new PulsarClientException.InvalidMessageException("Unexpected null message value"));
+                            throw elasticsearchClient.irrecoverableError.get();
+                    }
+                } else {
+                    if (elasticSearchConfig.isBulkEnabled()) {
+                        elasticsearchClient.bulkIndex(record, idAndDoc);
+                    } else {
+                        elasticsearchClient.indexDocument(record, idAndDoc);
+                    }
+                }
+            } catch (Exception e) {
```

Comment:
can we catch a more specific set of Exceptions ?

Reply:
This is to trace all unexpected Exception.

Reply:
in theory the exception should bubble up.

if you catch generic Exception then you have to deal with system exceptions.

at least we have to deal with InterruptedException as a special case and set `Thread.currentThread().interrupted()`

Reply:
It catches and re-throw Exception, as the Sink.write() can throw Exception. What's wrong here ?